src/qutip_cupy/expectation.py

```diff
@@ -0,0 +1,71 @@
+import cupy as cp
+
+
+def expect_cupydense(op, state):
+    """
+    Get the expectation value of the operator `op` over the state `state`.  The
+    state can be either a ket or a density matrix.
+
+    The expectation of a state is defined as the operation:
+        state.adjoint() @ op @ state
+    and of a density matrix:
+        tr(op @ state)
+    """
+    if state.shape[1] == 1:
+        return _expect_dense_ket(op, state)
+    return _expect_dense_dense_dm(op, state)
+
+
+def _expect_dense_ket(op, state):
+    _check_shape_ket(op, state)
+    return cp.vdot(state._cp, op._cp @ state._cp).item()
+
+
+def _check_shape_dm(op, state):
+    if (
+        op.shape[1] != state.shape[0]  # Matrix multiplication
+        or state.shape[0] != state.shape[1]  # State is square
+        or op.shape[0] != op.shape[1]  # Op is square
+    ):
+        raise ValueError(
+            "incorrect input shapes " + str(op.shape) + " and " + str(state.shape)
+        )
+
+
+_expect_dense_kernel = cp.RawKernel(
+    r"""
+    #include <cupy/complex.cuh>
+    extern "C" __global__
+    void expect_dens(const complex<double>* x1,const complex<double>* x2,
```

Comment:
Could we name this `expect_dense` and rename `x1` to `op` and `x2` to `dm` or `state` to make everything clearer?